go/payload-commit/worker/worker.go

```diff
@@ -0,0 +1,123 @@
+package worker
+
+import (
+	"context"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/remeh/sizedwaitgroup"
+	log "github.com/sirupsen/logrus"
+	"github.com/swagftw/gi"
+
+	"audit-protocol/goutils/settings"
+	"audit-protocol/goutils/taskmgr"
+	rabbitmqMgr "audit-protocol/goutils/taskmgr/rabbitmq"
+	worker2 "audit-protocol/goutils/taskmgr/worker"
+	"audit-protocol/payload-commit/service"
+)
+
+type Worker struct {
+	service  *service.PayloadCommitService
+	taskmgr  taskmgr.TaskMgr
+	settings *settings.SettingsObj
+}
+
+func (w *Worker) ConsumeTask() error {
+	// create buffered channel to limit the number of concurrent tasks.
+	// buffered channel can is used to accept multiple messages and then process them in parallel.
+	// as messages are generated at lower pace than they are consumed, we can use unbuffered channel as well.
+	// TBD: we can use unbuffered channel as well.
+	taskChan := make(chan taskmgr.TaskHandler, 5)
+	defer close(taskChan)
+
+	// start consuming messages in separate go routine.
+	// messages will be sent back over taskChan.
+	go func() {
+		err := backoff.Retry(func() error {
+			err := w.taskmgr.Consume(context.Background(), worker2.TypePayloadCommitWorker, taskChan)
+			if err != nil {
+				log.WithError(err).Error("failed to consume the message, retrying")
+
+				return err
+			}
+
+			return nil
+		}, backoff.NewExponentialBackOff())
+
+		if err != nil {
+			log.WithError(err).Fatal("failed to consume the messages after max retries")
+		}
+	}()
+
+	// create a wait group to wait for previous the tasks to finish.
+	// limit number of concurrent tasks per Worker
+	swg := sizedwaitgroup.New(5)
+
+	for {
+		swg.Add()
+
+		// as task chan is buffered channel, we can use it as a semaphore to limit the number of concurrent tasks.
+		taskHandler := <-taskChan
+
+		go func(taskHandler taskmgr.TaskHandler) {
+			msgBody := taskHandler.GetBody()
+
+			log.WithField("msg", string(msgBody)).Debug("received message")
+
+			err := w.service.Run(msgBody, taskHandler.GetTopic())
+			if err != nil {
+				log.WithError(err).Error("failed to run the task")
+
+				err = backoff.Retry(func() error {
+					return taskHandler.Nack(false)
+				}, backoff.NewExponentialBackOff())
+				if err != nil {
+					log.WithError(err).Errorf("failed to nack the message")
+				}
+			} else {
+				err = backoff.Retry(func() error {
+					return taskHandler.Ack()
+				}, backoff.NewExponentialBackOff())
+				if err != nil {
+					log.WithError(err).Error("failed to ack the message")
+				}
+			}
+
+			swg.Done()
+		}(taskHandler)
+
+		// wait till all the previous tasks are finished.
+		swg.Wait()
+	}
+}
+
+// should be implemented by all the workers
+var _ worker2.Worker = (*Worker)(nil)
```

Comment:
let's try to use better variable names, worker2 doesn't convey much information about what it is for.